### PR TITLE
Sorted the string.fsi entries and added examples to each one

### PR DESCRIPTION
**Pull request by MecuStefan**

### CODE REVIEW

**Comment by dnfadmin** [comment]
[CLA assistant check](https://cla.dotnetfoundation.org/dotnet/fsharp?pullRequest=12130) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/fsharp?pullRequest=12130) before we can accept your contribution.<br/><br/>:x: MecuStefan [sign now](https://cla.dotnetfoundation.org/dotnet/fsharp?pullRequest=12130)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/fsharp?pullRequest=12130) it.</sub>

**Comment by baronfel** [comment]
Adding a link to the [meta-issue](https://github.com/dotnet/fsharp/issues/12124) for tracking. Thanks for submitting these!

File: src/fsharp/FSharp.Core/string.fsi [hunk 1]
```diff
@@ -18,6 +18,29 @@ namespace Microsoft.FSharp.Core
...
+        /// strings.</summary>
+        ///
+        /// <param name="mapping">The function to produce a string from each character of the input string.</param>
+        /// <param name="str">The input string.</param>
+        ///
+        /// <returns>The concatenated string.</returns>
+        /// 
+        /// <example> The following samples shows how to interspace spaces in a text
+        /// <code lang="fsharp">
+        /// String.collect (sprintf "%c ") "Stefan says: Hi!"  // evaluates "S t e f a n   s a y s :   H i ! "
```
Comment (dsyme) [inline src/fsharp/FSharp.Core/string.fsi:32]: Use pipelining for each example, and suggest binding the input to make the example stand out more prominently:
```fsharp
        /// let inp = "Stefan says: Hi!"
        /// inp |> String.collect (sprintf "%c ") // evaluates "S t e f a n   s a y s :   H i ! "
```

File: src/fsharp/FSharp.Core/string.fsi [hunk 2]
```diff
@@ -27,54 +50,32 @@ namespace Microsoft.FSharp.Core
         /// <returns>A new string consisting of the concatenated strings separated by
         /// the separation string.</returns>
         /// <exception cref="T:System.ArgumentNullException">Thrown when <c>strings</c> is null.</exception>
+        /// 
+        /// <example>
+        /// <code lang="fsharp">
+        /// String.concat " " ["Stefan"; "says:"; "Hello"; "there!"]  // evaluates "Stefan says: Hello there!"
+        /// [0..9] |> List.map string |> String.concat ""             // evaluates "0123456789"
+        /// String.concat "!" ["No exclamation point here"]           // evaluates "No exclamation point here"
+        /// </code>
```
Comment (dsyme) [inline src/fsharp/FSharp.Core/string.fsi:59]: Suggest this:
```fsharp
        /// let inp1 = ["Stefan"; "says:"; "Hello"; "there!"]
        /// inp1 |> String.concat " "   // evaluates "Stefan says: Hello there!"
        ///
        /// let inp2 = [0..9]
        /// [0..9] |> List.map string |> String.concat ","             // evaluates "0,1,2,3,4,5,6,7,8,9"
        ///
        /// let inp3 = "No comma"
        /// inp3 |> String.concat ","            // evaluates "No comma"
```

File: src/fsharp/FSharp.Core/string.fsi [hunk 3]
```diff
@@ -27,54 +50,32 @@ namespace Microsoft.FSharp.Core
...
-        /// <returns>The concatenated string.</returns>
-        [<CompiledName("Collect")>]
-        val collect: mapping:(char -> string) -> str:string -> string
+        /// <returns>True if any character returns true for the predicate and false otherwise.</returns>
+        /// 
+        /// <example> Looking for uppercase characters
+        /// <code lang="fsharp">
+        /// String.exists System.Char.IsUpper "Yoda"  // evaluates true
+        /// String.exists System.Char.IsUpper "nope"  // evaluates false
+        /// </code>
```
Comment (dsyme) [inline src/fsharp/FSharp.Core/string.fsi:75]: Suggest opening System and using pipeline:
```fsharp
        /// open System
        /// "Yoda" |> String.exists Char.IsUpper     // evaluates true
        /// "nope" |> String.exists Char.IsUpper    // evaluates false
```

File: src/fsharp/FSharp.Core/string.fsi [hunk 4]
```diff
@@ -85,9 +86,32 @@ namespace Microsoft.FSharp.Core
         /// <param name="str">The input string.</param>
         ///
         /// <returns>The resulting string.</returns>
+        /// 
+        /// <example> Filtering out just alphanumeric characters or just digits
+        /// <code lang="fsharp">
+        /// String.filter System.Uri.IsHexDigit "0 1 2 3 4 5 6 7 8 9 a A m M"  // evaluates "123456789aA"
+        /// String.filter System.Char.IsDigit "hello"                          // evaluates ""
+        /// </code>
```
Comment (dsyme) [inline src/fsharp/FSharp.Core/string.fsi:94]: Similar to comments above 
* separate the examples by blank lines
* if any of the inputs are longer then ~5chars, then bind each of the inputs to `inp1` etc 
* open `System` if needed
* use pipeline

File: src/fsharp/FSharp.Core/string.fsi [hunk 5]
```diff
@@ -98,26 +122,115 @@ namespace Microsoft.FSharp.Core
...
+        /// 
+        /// <example>
+        /// <code lang="fsharp">
+        /// String.length null   // evaluates 0
+        /// String.length ""     // evaluates 0
+        /// String.length "123"  // evaluates 3
+        /// </code>
+        /// </example>
+        /// 
+        /// <example>
```
Comment (dsyme) [inline src/fsharp/FSharp.Core/string.fsi:168]: This is good.  Separating into multiple examples is totally fine. Do the same for those further above if you think it's right

File: src/fsharp/FSharp.Core/string.fsi [hunk 6]
```diff
@@ -18,6 +18,29 @@ namespace Microsoft.FSharp.Core
...
+        /// <summary>Builds a new string whose characters are the results of applying the function <c>mapping</c>
+        /// to each of the characters of the input string and concatenating the resulting
+        /// strings.</summary>
+        ///
+        /// <param name="mapping">The function to produce a string from each character of the input string.</param>
+        /// <param name="str">The input string.</param>
+        ///
+        /// <returns>The concatenated string.</returns>
+        /// 
+        /// <example> The following samples shows how to interspace spaces in a text
```
Comment (baronfel) [inline src/fsharp/FSharp.Core/string.fsi:30]: This is actually the first time I've seen multiple examples in the wild, so congratulations on surprising me :)

I did check, and FSharp.Formatting and FsAutoComplete _do_ handle multiple examples just fine, so this shouldn't be a problem from the perspective of tooling consuming these docs.

@dsyme it does make me want some kind of 'example identifier' for use in generation, for example generating a link of `FSharp.Core.StringModule#collect-example-collect-space` in order to jump to the specific example, but that's just some off-the-cuff musing.

**Review by dsyme — CHANGES_REQUESTED** [review]
This is a fantastic start - the examples are really well chosen.

Have left some comments - could you aplpy them systematically and I'll review again ?

File: src/fsharp/FSharp.Core/string.fsi (hunk 6 above)
Comment (dsyme) [reply src/fsharp/FSharp.Core/string.fsi:30]: > @dsyme it does make me want some kind of 'example identifier' for use in generation, for example generating a link of FSharp.Core.StringModule#collect-example-collect-space in order to jump to the specific example, but that's just some off-the-cuff musing. (FSharp.Formatting already generates a numeric identifier for this case, but something more stable might be useful?)

Makes sense, yes, even just `collect-example-1`

Comment (dsyme) [reply src/fsharp/FSharp.Core/string.fsi:30]: just use `<example id="collect-example-whatever">` I think

Comment (dsyme) [reply src/fsharp/FSharp.Core/string.fsi:30]: > I did check, and FSharp.Formatting and FsAutoComplete do handle multiple examples just fine, so this shouldn't be a problem from the perspective of tooling consuming these docs.

Thank you for checking this!!!!

Comment (baronfel) [reply src/fsharp/FSharp.Core/string.fsi:30]: yeah, that's what I was getting at, I just wonder if it's worth codifying that in the tooling RFC and adding support to FSharp.Formatting then.

Comment (dsyme) [reply src/fsharp/FSharp.Core/string.fsi:30]: > yeah, that's what I was getting at, I just wonder if it's worth codifying that in the tooling RFC and adding support to FSharp.Formatting then.

Absolutely

Comment (baronfel) [reply src/fsharp/FSharp.Core/string.fsi:30]: tracking this [on the RFC](https://github.com/fsharp/fslang-design/pull/630) and in [FSharp.Formatting](https://github.com/fsprojects/FSharp.Formatting/pull/704)

Comment (MecuStefan) [reply src/fsharp/FSharp.Core/string.fsi]: Added ids to every example tag

Comment (MecuStefan) [reply src/fsharp/FSharp.Core/string.fsi]: Facepalm you want to generate the ids ...

Comment (baronfel) [reply src/fsharp/FSharp.Core/string.fsi]: no, I'd keep them. My intent was for the author to give them meaningful ids. There is some last-chance id generation happening in FSharp.Formatting, but the MR I have up now always takes the `id` that is present if there is one.  So, all in all, I think what you have now is great 👍

File: src/fsharp/FSharp.Core/string.fsi [hunk 7]
```diff
@@ -98,26 +144,115 @@ namespace Microsoft.FSharp.Core
...
-        val forall: predicate:(char -> bool) -> str:string -> bool
+        /// <param name="action">The function to apply to each character and index of the string.</param>
+        /// <param name="str">The input string.</param>
+        /// 
+        /// <example id="iteri-example-1"> Numbering the characters and printing the associated ASCII code 
+        /// for each characater in the input string
+        /// <code lang="fsharp">
+        /// let input = "Hello"
+        /// input |> String.iteri (fun i c -> printfn "%d. %c %d" (i + 1) c (int c))
+        /// // evaluates unit
```
Comment (dsyme) [inline src/fsharp/FSharp.Core/string.fsi:188]: I think these english language parts should not be in `<code>`

File: src/fsharp/FSharp.Core/string.fsi [hunk 8]
```diff
@@ -98,26 +144,115 @@ namespace Microsoft.FSharp.Core
         ///
         /// <returns>The constructed string.</returns>
         /// <exception cref="T:System.ArgumentException">Thrown when <c>count</c> is negative.</exception>
+        /// 
+        /// <example id="init-example-1"> Enumerate digits ASCII codes
+        /// <code lang="fsharp">
+        /// String.init 10 (fun i -> int '0' + i |> sprintf "%d ")  // evaluates "48 49 50 51 52 53 54 55 56 57 "
```
Comment (dsyme) [inline src/fsharp/FSharp.Core/string.fsi:150]: These lines are getting too long I think (102 characters), suggest a maximum of 80 with a preference for 60
```fsharp
        /// <code lang="fsharp">
        /// String.init 10 (fun i -> int '0' + i |> sprintf "%d ")
        /// </code>
        /// evaluates to
        /// <code>
        /// "48 49 50 51 52 53 54 55 56 57 "
        /// </code>
```

across all the samples.

File: src/fsharp/FSharp.Core/string.fsi [hunk 9]
```diff
@@ -98,26 +144,115 @@ namespace Microsoft.FSharp.Core
...
         ///
-        /// <param name="predicate">The function to test each character of the string.</param>
+        /// <param name="action">The function to be applied to each character of the string.</param>
         /// <param name="str">The input string.</param>
+        /// 
+        /// <example id="iter-example-1"> Printing the ASCII code for each characater in the string
+        /// <code lang="fsharp">
+        /// let input = "Hello"
+        /// input |> String.iter (fun c -> printfn "%c %d" c (int c))
+        /// // evaluates unit
```
Comment (dsyme) [inline src/fsharp/FSharp.Core/string.fsi:165]: English language parts outside of `<code>`

**Review by dsyme — APPROVED** [review]
I've left some comments/improvements - I'll merge and could you address them on next PR?  Thanks

**Pull request merged**